apple/internal/xctrunner.bzl

```diff
@@ -0,0 +1,157 @@
+    "//apple:providers.bzl",
+    "AppleBundleInfo",
+)
+load(
+    "//apple/internal:providers.bzl",
+    "new_applebinaryinfo",
+)
+
+_TestBundleInfo = provider(
+    "Test bundle info for tests that will be run.",
+    fields = {
+        "infoplists": "A `depset` of `File`s of `Info.plist` files.",
+        "xctests": "A `depset` of paths of XCTest bundles.",
+    },
+)
+
+def _test_bundle_info_aspect_impl(target, ctx):
+    rule_attr = ctx.rule.attr
+
+    if AppleBundleInfo in target:
+        info = target[AppleBundleInfo]
+        xctests = depset([info.archive])
+        infoplists = depset([info.infoplist])
+    else:
+        deps = getattr(rule_attr, "tests", [])
+        xctests = depset(
+            transitive = [
+                dep[_TestBundleInfo].xctests
+                for dep in deps
+            ],
+        )
+        infoplists = depset(
+            transitive = [
+                dep[_TestBundleInfo].infoplists
+                for dep in deps
+            ],
+        )
+
+    return [
+        _TestBundleInfo(
+            infoplists = infoplists,
+            xctests = xctests,
+        ),
+    ]
+
+test_bundle_info_aspect = aspect(
+    attr_aspects = ["tests"],
+    implementation = _test_bundle_info_aspect_impl,
+)
+
+def _xctrunner_impl(ctx):
+    # Get test target info
+    infos = [target[_TestBundleInfo] for target in ctx.attr.test_targets]
+    infoplists = depset(
+        transitive = [info.infoplists for info in infos],
+    )
+    xctests = depset(
+        transitive = [info.xctests for info in infos],
+    )
+    output = ctx.actions.declare_directory(ctx.label.name + ".app")
+
+    # Args for _make_xctrunner
+    arguments = ctx.actions.args()
+    arguments.add("--name", ctx.label.name)
+    arguments.add("--platform", ctx.attr.platform)
+
+    if ctx.attr.verbose:
+        arguments.add("--verbose", ctx.attr.verbose)
+
+    # Absolute paths to xctest bundles
+    arguments.add_all(
+        xctests,
+        before_each = "--xctest",
+        expand_directories = False,
+    )
+
+    # App bundle output path
+    arguments.add("--output", output.path)
+
+    ctx.actions.run(
+        inputs = depset(transitive = [xctests, infoplists]),
+        outputs = [output],
+        executable = ctx.executable._make_xctrunner,
+        arguments = [arguments],
+        mnemonic = "MakeXCTRunner",
+    )
+
+    # Limiting the contents of AppleBinaryInfo to what is necessary
+    # for testing and validation.
+    xctrunner_binary_info = new_applebinaryinfo(
+        binary = output,
+        infoplist = None,
+        product_type = None,
+    )
+
+    return [
+        DefaultInfo(files = depset([output])),
+        xctrunner_binary_info,
+    ]
+
+xctrunner = rule(
+    implementation = _xctrunner_impl,
+    attrs = {
+        "platform": attr.string(
+            default = "iPhoneOS.platform",
+            mandatory = False,
+            doc = "Platform to bundle for. Default: iPhoneOS.platform",
+        ),
```

Comment:
We should be able to collect this via the aspect, and error in analysis if the platforms don't match between the tests.